### PR TITLE
docs: add how using `ResponseTrait` in custom classes

### PR DESCRIPTION
**Pull request by datamweb**
**Description**
In the current CodeIgniter documentation, the usage of ResponseTrait is primarily explained in the context of controllers. However, there are scenarios where developers might want to use this trait in their custom classes that are not controllers. Unfortunately, there is no documentation on how to achieve this. This PR adds a new section to the CodeIgniter documentation explaining how to utilize ResponseTrait in custom classes.

When developers try to use ResponseTrait in custom classes that do not extend from Controller, they encounter the error Undefined property: $response. This happens because ResponseTrait relies on properties that are typically available in controllers, and without manually setting up a Response object in custom classes, the trait cannot function properly.

This PR provides guidance on how to manually create an instance of the Response class within custom classes and make it accessible to ResponseTrait. By doing so, developers can leverage the capabilities of ResponseTrait in non-controller classes.

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by datamweb** [comment]
Traits in PHP are a powerful feature that allows for the **reuse of methods across multiple classes without requiring inheritance**. 

**Key characteristics of traits include:**

- Reusability Across Different Classes: Traits enable you to define methods that can be reused in various classes, regardless of their inheritance hierarchy. This helps in avoiding code duplication and promoting code reuse.
- Decoupling Functionality from Inheritance: Unlike traditional inheritance, where functionality is inherited through class hierarchies, traits allow developers to add methods to classes without imposing any inheritance constraints. This allows classes to maintain their primary inheritance while gaining additional behavior from traits.
- Method Composition: Traits support method composition by allowing classes to include multiple traits, and traits can also include other traits. This capability makes traits a powerful tool for combining complex behaviors without rigid class structures.
- Conflict Resolution: PHP provides mechanisms for resolving conflicts when multiple traits define methods with the same name, allowing developers to specify which method should take precedence.


Currently, in CodeIgniter, traits like ResponseTrait are tightly coupled with the Controller class. This design enforces that ResponseTrait depends on properties such as `$response`, which are automatically available in controllers but not in custom classes like libraries, models, or services. This limitation restricts the flexibility of ResponseTrait, contradicting the core purpose of traits in PHP.


**This limitation can be considered a bug from a design perspective. Traits are intended to be reusable across various classes without being tied to a specific class structure. The current implementation, which restricts ResponseTrait to controllers, undermines the flexibility and modularity that traits are supposed to offer.**

**Proposed Solution**
To address this issue, I propose a simple refactor that would enhance ResponseTrait by using `service('response')` internally. This change would remove the dependency on the `$response` property, allowing ResponseTrait to be used in any class.

**Comment by kenjis** [comment]
This is a topic that needs to be carefully considered, because traits are easy to use anywhere, but in reality they are very difficult to use.

References:
- The Fallacy of Overreliance on Traits in PHP: Unmasking Poor Class Design | PullRequest Blog https://www.pullrequest.com/blog/the-fallacy-of-overreliance-on-traits-in-php-unmasking-poor-class-design/
- When to use a trait? — Matthias Noback - Blog https://matthiasnoback.nl/2022/07/when-to-use-a-trait/
- PHP Traits – The Good, the Bad and the Ugly – Vasanth Krishnamoorthy https://www.vasanthk.com/php-traits-the-good-the-bad-and-the-ugly/

**Comment by kenjis** [comment]
> Currently, in CodeIgniter, traits like ResponseTrait are tightly coupled with the Controller class. This design enforces that ResponseTrait depends on properties such as $response, which are automatically available in controllers but not in custom classes like libraries, models, or services.

Yes. According to the user guide, API Response trait is only for controllers.

> CodeIgniter provides an API Response trait that **can be used with any controller** to make common response types simple, with no need to remember which HTTP status code should be returned for which response types.
> https://codeigniter4.github.io/CodeIgniter4/outgoing/api_responses.html

And the ResponseTrait provides methods to return Response objects.
Response objects are objects in the HTTP layer.
In CI4, only controllers and controller filters are the parts that return Response objects.

Models should clearly be unrelated to the HTTP layer. Models handle the database layer (or business logic). 
So it should not touch Responses.

Personally, I think that everything related to HTTP should be handled within the controller (or filter).
So I don't know there are scenarios where you may want to use this trait in custom classes that do not extend from the base Controller class.

**Comment by datamweb** [comment]
> So I don't know there are scenarios where you may want to use this trait in custom classes that do not extend from the base Controller class.

**Using PaymentService Across Multiple Controllers**

The PaymentService handles the complex logic of processing payments and uses the ResponseTrait to generate appropriate HTTP responses. This service can easily be utilized by any controller that requires payment processing.

```php
<?php
use CodeIgniter\API\ResponseTrait;

class PaymentService
{
    use ResponseTrait;

    public function processPayment($orderId, $paymentDetails)
    {
        try {
            // 1. Validate payment details
            if (!$this->validatePaymentDetails($paymentDetails)) {
                return $this->failValidationError('Invalid payment details.');
            }

            // 2. Process transaction with external API
            $transactionResult = $this->processTransactionWithGateway($orderId, $paymentDetails);

            if (!$transactionResult->isSuccessful()) {
                return $this->fail('Payment processing failed.');
            }

            // 3. Update order status
            $this->updateOrderStatus($orderId, 'paid');

            // 4. Send successful response
            return $this->respondCreated('Payment processed successfully.');
        } catch (\Exception $e) {
            return $this->failServerError('An unexpected error occurred during payment processing.');
        }
    }

    protected function validatePaymentDetails($paymentDetails)
    {
        // Logic to validate payment details
        return true;
    }

    protected function processTransactionWithGateway($orderId, $paymentDetails)
    {
        // Logic to interact with the payment gateway
        // Assume external API interaction happens here
        return (object) ['isSuccessful' => true];
    }

    protected function updateOrderStatus($orderId, $status)
    {
        // Update order status in the database
    }
}
```

2. Cart Controller (CartController):
This controller is responsible for handling user purchases and uses PaymentService during the checkout process.

```php
<?php
use CodeIgniter\Controller;

class CartController extends Controller
{
    protected $paymentService;

    public function __construct()
    {
        $this->paymentService = new PaymentService();
    }

    public function checkout()
    {
        $orderId = 123; // Assume this is provided
        $paymentDetails = $this->request->getPost('payment');

        // Use PaymentService to process the payment
        $response = $this->paymentService->processPayment($orderId, $paymentDetails);
        
        // Return the response to the user
        return $response;
    }
}
```
3. Order Management Controller (OrderController):
This controller is used to manage orders and may need to confirm or reprocess a payment for a specific order.

```php
<?php
use CodeIgniter\Controller;

class OrderController extends Controller
{
    protected $paymentService;

    public function __construct()
    {
        $this->paymentService = new PaymentService();
    }

    public function retryPayment($orderId)
    {
        $paymentDetails = $this->request->getPost('payment');

        // Use PaymentService to reprocess the payment
        $response = $this->paymentService->processPayment($orderId, $paymentDetails);
        
        // Return the response to the user
        return $response;
    }
}
```
4. Mobile API Controller (MobileApiController):
This controller handles API requests from a mobile app where users can make payments.

```php
<?php
use CodeIgniter\Controller;

class MobileApiController extends Controller
{
    protected $paymentService;

    public function __construct()
    {
        $this->paymentService = new PaymentService();
    }

    public function mobilePayment()
    {
        $orderId = $this->request->getPost('orderId');
        $paymentDetails = $this->request->getPost('payment');

        // Use PaymentService to process the payment via the mobile API
        $response = $this->paymentService->processPayment($orderId, $paymentDetails);
        
        // Return the response to the user
        return $response;
    }
}
```
Benefits of This Architecture:
Separation of Concerns: The payment processing logic is separated from the controllers and centralized in the PaymentService. This allows controllers to focus only on handling HTTP requests.

Reusability: The PaymentService can be reused across multiple controllers or other parts of the system, reducing code duplication and enhancing maintainability.

Simpler Controllers: Controllers only serve as a bridge between HTTP requests and services, reducing their complexity and improving readability.

Better Testability: The service can be easily unit tested in isolation from HTTP context, making it easier to verify the business logic without mocking HTTP requests.

**Comment by datamweb** [comment]
Your feedback is really helpful and will definitely contribute to making the code cleaner and more maintainable. However, I do have a few different perspectives on some of the points you raised, and I think these could be explored from another angle.

### 1. **Over-splitting of Classes and the Violation of YAGNI (You Aren't Gonna Need It)**
Regarding the suggestion to split the `PaymentService` into multiple classes (like `PaymentProcessor`, `OrderService`, and `ResponseService`), this might not always be necessary, especially in smaller or mid-sized projects. According to the YAGNI principle, we shouldn't implement something prematurely unless we have a real need for it. Over-architecting early on can increase code complexity and maintenance costs without clear immediate benefits. This separation could be done progressively as the project grows.

### 2. **Dependency on ResponseTrait and Full Decoupling from HTTP**
You mentioned that using `ResponseTrait` in the `PaymentService` tightly couples the class to HTTP, which is a valid concern. However, in cases where the class is specifically designed to respond to HTTP requests, this dependency might actually enhance simplicity and readability. If the main purpose of the class is to work within the HTTP context, maintaining this coupling isn’t necessarily a bad thing and may avoid unnecessary abstractions or complex dependency injections.

### 3. **Avoiding Dependency Injection in Small Projects**
While Dependency Injection (DI) is indeed a great practice for large and complex projects, in some smaller or mid-sized projects, it might not be necessary to fully implement DI. Direct instantiation using `new` can be a simpler and more pragmatic choice, especially when the number of services being injected is low. Full DI implementation might introduce extra overhead in situations where it's not entirely justified.

### 4. **Using Exceptions vs. Localized Error Handling**
You suggested using exceptions for error handling, which is a valid approach. However, in some cases, local error handling within the methods themselves can be just as acceptable, particularly when immediate error reporting is required. Using exceptions moves error handling up the stack, which can add complexity if exceptions need to be managed in various places. In some scenarios, keeping error management localized might be simpler and more effective.

In conclusion, I’m not entirely against your suggestions; in fact, I think your design proposals are quite logical and applicable in many scenarios. However, the points I raised could also serve as valid alternatives in specific situations. I believe that diversity of perspectives and differences in design opinions can drive a project forward and help us discover the best solutions.

Thanks!

**Comment by kenjis** [comment]
Yes! Any issues with the software design are “dependent on the situation”.

The object-oriented design principle is ultimately to be able to respond to future changes in functionality without changing the existing code as much as possible.

If a class is not changed in the future, it is complete and there are no problems. 
On the other hand, if the class has a part that is changed, it should be designed so that another object can be injected .
